lib/ecmascript.ts

```diff
@@ -567,13 +567,29 @@ export function ParseTemporalDurationString(isoString: string) {
 
 // ts-prune-ignore-next TODO: remove if test/validStrings is converted to TS.
 export function ParseTemporalInstant(isoString: string) {
-  const { year, month, day, hour, minute, second, millisecond, microsecond, nanosecond, offset, z } =
+  let { year, month, day, hour, minute, second, millisecond, microsecond, nanosecond, offset, z } =
     ParseTemporalInstantString(isoString);
 
+  if (!z && !offset) throw new RangeError('Temporal.Instant requires a time zone offset');
+  // At least one of z or offset is defined, but TS doesn't seem to understand
+  // that we only use offset if z is not defined (and thus offset must be defined).
+  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
```

Comment:
I don't think typescript's inference is good enough with destructuring of those sorts of unions to enable better type-checking than this. If we rewrote this code to avoid destructuring the returned object then it might stand a chance. :)